Tensorflow minimum version set to 2.14.0 + removing napari and optional cupy dependency from toml + improvements to readme

ShantanuKodgirwar opened the pull request:
## PR changes:
- Higher flexibility for `tensorflow` as the previous version was constrained to 2.13.0 which caused issues with transitive dependencies such as `typing_instructions`.
- Removing the optional dependency for `cupy` and instead redirecting user to install it from their documentation
- The `example_scripts/exampleReconstructionCPM.py` file is modified slightly to use `argparse` for CLI utility and some other minor things.
- Modified `README.md`:
   - includes simplified explanation for installation (removing the optional flags)
   - Directing the user to install `cupy` manually (no more optional GPU flags) as it is device dependent and would allow them to choose how to install it.
   - details on how to download and run reconstruction directly as a CLI command, which can be nicer for a new user before looking into examples.
   - Changing the package management instructions in the end which can use `conda` and `poetry` together.
- Minor changes to `.gitignore`

### EDIT:

I made additional changes in the mean time which I think are also important. These changes are also written below, but copying here for quick reference.
- removing napari as a dependecy because it is not used for core implementation. It relies on a compulsory interactive backend (which may not be possible to use on a remote server without X11 port forwarding). Moreover, napari is a heavy dependency and increases package installation time. 
- Some non-used napari-based implementations have a separate and rather "hacky" try-except block that checks if napari is installed and if someone wants to test these non-core implementations, they can install napari manually on their own. Perhaps eventually these non-core implementations could be moved to a separate optional sub-package.
- Minor changes to readme for the package management section
- `pyqt5` as a dependency is added again as without this dependency `pyqtgraph` a compulsory dependency does not work and causes error when PtyLab is imported (sorry I missed this critical issue before!).
- When running on CPU, the message "cupy unavailable, cannot use GPU for faster reconstruction." seems to be imported from many files and prints multiple times (a little unnecessary). This message is commented out from these files - black formatted the files where the print message was commented.
- Now checks the availability of cupy (and cuda as well) just once in the `__init__.py` file. 
- Fixing legacy imports `fracPy` from `Engines/e3PIE_fixed.py`. Although not sure if this implementation is production ready or some redundant implementation of `e3PIE.py`.
- Numpy constraint ">=1.22, <2.0.0". Let's keep this flexible only till 2.0.0. The new major numpy release 2.0.0 has a lot of breaking changes.
- Bump `ptylab` to minor release 0.2.0

@dboonz @roflmaostc If these seem to be a lot of changes, we could create a minor release to 0.2.0 as well.

ShantanuKodgirwar commented:
@roflmaostc @dboonz

ShantanuKodgirwar commented on pyproject.toml:
```diff
@@ -1,6 +1,6 @@
+version = "0.1.2"
 description = "A cross-platform, open-source inverse modeling toolbox for conventional and Fourier ptychography"
 authors = ["Lars Loetgering <lars.loetgering@fulbrightmail.org>", "PtyLab Team"]
 license = "MIT"
```
True. I'll remove that.

ShantanuKodgirwar commented on pyproject.toml:
```diff
@@ -1,6 +1,6 @@
 [tool.poetry]
 name = "ptylab"
-version = "0.1.1"
+version = "0.1.2"
```
There is no breaking change from within the package itself. I'd stick with 0.1.2

ShantanuKodgirwar replied:
Although one could argue that the optional flags (`gpu11x` and `gpu12x`) are not available anymore and the previous installation with `pip install .[gpu11x]` won't work again. But that is not a breaking change from the code base itself. So it makes sense to do a patch release than a minor release.

dboonz commented:
Looks like you did a lot of work! I'll take a look on Monday and get back
to you.

On Wed, Jun 5, 2024, 10:59 Shantanu Kodgirwar ***@***.***>
wrote:

> @roflmaostc <https://github.com/roflmaostc> @dboonz
> <https://github.com/dboonz>
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/PtyLab/PtyLab.py/pull/58#issuecomment-2149261992>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAR23V4YV62DLFRTTXVIL4DZF3HQFAVCNFSM6AAAAABIMUCWY6VHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDCNBZGI3DCOJZGI>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

ShantanuKodgirwar commented:
I think currently the library is very heavy to install. In the end with some cleanups, tensorflow, napari and bokeh can be made as optional dependencies because the core implementation does not really use these dependencies. The files using them could be added to some sort of optional modules perhaps in the next 0.2.0 version.

roflmaostc commented:
Any news @dboonz ?

ShantanuKodgirwar commented:
New changes:
- removing napari as a dependecy because it is not used for core implementation. It relies on a compulsory interactive backend (which may not be available on a remote server or would require a separate installation). Moreover, napari is a heavy dependency and increases package installation time. 
- Some non-used napari-based implementations have a separate try-except block that checks if napari is installed and if someone wants to test these non-core implementations, they can install napari manually on their own.

ShantanuKodgirwar commented:
New changes:
- `pyqt5` as a dependency is added again as without this dependency `pyqtgraph` a compulsory dependency does not work and causes error when PtyLab is imported (sorry I missed this critical issue before!).
- When running on CPU, the message "cupy unavailable, cannot use GPU for faster reconstruction." seems to be imported from many files and prints multiple times (a little unnecessary). This message is commented out from these files + black formatted these files 
- Now checks the availability of cupy and cuda as well just once in the `__init__.py` file.

ShantanuKodgirwar commented:
New changes:

Synced the main branch and resolved merge conflicts. Plus again tested, installation, reconstruction and proper imports.

roflmaostc commented:
Maybe @dboonz can have a final look?

dboonz commented:
Hey! I took a look, thanks a lot for the hard work!

I added a little explanation to the exampleCPM.py, as we may have some beginner programmers who have no idea what argparse actually is.

Also, it seems that if we run the file now with the example data generated by simulateData.py, the reconstruction fails. That's a bit too realistic ;). Could you check if it's just me or that you run into the same problem? It's the first thing people will probably try out..

ShantanuKodgirwar commented:
Hi Dirk,

You're welcome. I do this anyways because I use ptylab. And thanks for making this package :)

> Also, it seems that if we run the file now with the example data generated by simulateData.py, the reconstruction fails.

Really? I just tried `python example_scripts/exampleReconstructionCPM.py --f example_data/simu.hdf5` and it works just fine for me. 

Btw I see that you added the tensorboard monitor as an optional monitor in the example, however, maybe it's a good idea to keep the default monitor as the one for the first run and if someone wants they can try the tensorboard monitor as well (added that in the latest commit). Also I am not very sure what am I looking at exactly. The probe and object look very non-standard to me. To me the `jupyter_tutorials/jupyter_tutorials_tutorial_CPM.ipynb` feels easier to see what is happening.

ShantanuKodgirwar commented:
@dboonz @roflmaostc 

To make it even easier to start off, I have added a colab link (with a colab badge) of the demo notebook (in the root directory). This colab demo is self-sufficient to get started as the first cell also installs the dependencies. Please note this colab link in readme is based on the assumption that it is merged to the main branch. To check it out right now, it must be accessed from this branch. The temporary link for this would be: [colab demo](https://colab.research.google.com/github/ShantanuKodgirwar/PtyLab.py/blob/tensorflow-dep-2.14.0/demo.ipynb). Feel free to run all the cells and see the reconstructions. Even @roflmaostc can try it out as it does not require local installation.

I have also added a gif of the reconstruction in readme below this link (I suppose it is instantly nicer for somehow to see the capabilities of the package right away).  

I think it would be a good idea to merge this PR, might greatly improve the visibility and accessibility of the package :)

roflmaostc commented:
Yes looks much nicer!

Ready to merge?

ShantanuKodgirwar commented:
Yeah!

ShantanuKodgirwar commented:
@roflmaostc @dboonz 

I am currently using aPIE and zPIE for my work, but I have encountered some issues. aPIE relies on a legacy interpolation method that has been discontinued in scipy 1.14.0, and zPIE does not converge well in my experience and also according to several active PtyLab.py users from Jan Rothhardt's group. Additionally, zPIE uses a different frontend (bokeh) for plotting, which is inconsistent with the rest of the project that uses matplotlib. I plan to send a separate PR to address these issues. For now, I recommend merging this PR to avoid overloading it.

ShantanuKodgirwar commented:
I also remember about CI for this package, but I'll need to find extra time for it. I'll send for a PR for that eventually.
